### PR TITLE
🦺[Tech debt] Refactor Gilded Rose inventory app

### PR DESCRIPTION
**Pull request by kcw-grunt**
# Gilded Rose (Swift)

## Refactor Overview

Let us take a moment of silence for Leeroy who worked hard on the Gilded Rose inventory system yet met his demise in a fiery blaze of glory.  We will always remember him for his love of chicken🍗.

Reviewing the spec, it was clear there needed to be a way to capture items that had attributes beyond the standard yet the exceptions needed tests to also allow flexibilty in the custom items.  This following details some of the steps to acheive this.

## Items and SuperItems

As we avoid the ire of the orge the original class was untouched yet a superclass was created to cover the Conjured and Legendary items.

So, inventory is separated into 3 groups `Regular`, `Conjured` and `Legendary`

|Regular|Conjured|Legendary|
|:------:|:------:|:------:|
|Elixir of the Mongoose|Conjured Mana Cake|Sulfuras, Hand of Ragnaros| 
|+5 Dexterity Vest| | |
|Aged Brie| | |
|Backstage passes to a TAFKAL80ETC concert| | |

## Approach

The expections in the instructions from management suggest this style will persist so notes and separation of concerns were followed as best as possible

- Comments and explanations were added within reason
- Subgroups were updated separately
- Exceptions where handled based on the spec.
- Hidden expectations were assumed based on the legacy system

## Testing

A set of tests were written to cover the business logic but there was not enough time to capture all corner cases. There were some test cases that should be considered to more thouroughly verify the design. Some tests if more time was spent:

- Add tests to check the value boundaries
- Add tests to check the hidden value decrease after sellin < 0
- Add tests to test limits of large numbers to anticpate higher O(n) complexity
- Add more formatted outputs for the end user to improve readability

## Notes

Modifying line 44 in main.swift to:

``` print(superItem.details());```

will print all parameters to include item UUID.

### CODE REVIEW

File: swift/Sources/GildedRoseApp/main.swift [hunk 1]
```diff
@@ -1,30 +1,102 @@
...
+    SuperItem(name: "+5 Dexterity Vest", sellIn: 10, quality: 20, isConjured: false),
+    SuperItem(name: "Aged Brie", sellIn: 2, quality: 0, isConjured: false),
+    SuperItem(name: "Elixir of the Mongoose", sellIn: 5, quality: 7, isConjured: false),
+    SuperItem(name: "Sulfuras, Hand of Ragnaros", sellIn: 0, quality: 80, isConjured: false),
+    SuperItem(name: "Sulfuras, Hand of Ragnaros", sellIn: -1, quality: 80, isConjured: false),
+    SuperItem(name: "Backstage passes to a TAFKAL80ETC concert", sellIn: 15, quality: 20, isConjured: false),
+    SuperItem(name: "Backstage passes to a TAFKAL80ETC concert", sellIn: 10, quality: 49, isConjured: false),
+    SuperItem(name: "Backstage passes to a TAFKAL80ETC concert", sellIn: 5, quality: 49, isConjured: false),
+    
+    // Conjured Items
```
Comment (kcw-grunt) [inline swift/Sources/GildedRoseApp/main.swift:21]: Separated for readability

**Comment by kcw-grunt** [comment]
@cameroncooke @Jackstone92
Thanks. I had alot of fun  (maybe too much?) 
Please let me know when you would like to discuss further.
